src/protocols/Protocols.h

```diff
@@ -32,27 +32,56 @@ namespace Protocols {
 //
 // CHIP Protocol Ids (32-bits max)
 //
-
-enum CHIPProtocolId
+class Id
 {
-    // Common Protocols
-    //
-    // NOTE: Do not attempt to allocate these values yourself.
-
-    kProtocol_SecureChannel       = (kChipVendor_Common << 16) | 0x0000, // Secure Channel Protocol
-    kProtocol_Echo                = (kChipVendor_Common << 16) | 0x0002, // Echo Protocol
-    kProtocol_BDX                 = (kChipVendor_Common << 16) | 0x0003, // Bulk Data Exchange Protocol
-    kProtocol_NetworkProvisioning = (kChipVendor_Common << 16) | 0x0004, // Network Provisioning Protocol
-    kProtocol_InteractionModel    = (kChipVendor_Common << 16) | 0x0005, // Interaction Model Protocol
-    kProtocol_FabricProvisioning  = (kChipVendor_Common << 16) | 0x0006, // Fabric Provisioning Protocol
-    kProtocol_ServiceProvisioning = (kChipVendor_Common << 16) | 0x0007, // Service Provisioning Protocol
-    kProtocol_OpCredentials       = (kChipVendor_Common << 16) | 0x0008, // Operational Credentials
-
-    // Protocols reserved for internal protocol use
-
-    kProtocol_NotSpecified = (kChipVendor_NotSpecified << 16) | 0xFFFF, // The profile ID is either not specified or a wildcard
+public:
+    constexpr Id(VendorId aVendorId, uint16_t aProtocolId) : mVendorId(aVendorId), mProtocolId(aProtocolId) {}
+
+    constexpr bool operator==(const Id & aOther) { return mVendorId == aOther.mVendorId && mProtocolId == aOther.mProtocolId; }
+
+    // Convert the Protocols::Id to a TLV profile id.
+    // NOTE: We may want to change the TLV reader/writer to take Protocols::Id
+    // directly later on and get rid of this method.
+    constexpr uint32_t ToTLVProfileId() const { return ToUint32(); }
+
+    // Convert the protocol id to a 32-bit unsigned integer "fully qualified"
+    // form as defined in the spec.  This should only be used in the places
+    // where the spec defines a 32-bit unsigned in as a wire representation of
+    // protocol id.
+    constexpr uint32_t ToFullyQualifiedSpecForm() const { return ToUint32(); }
+
+    constexpr VendorId GetVendorId() const { return mVendorId; }
+    constexpr uint16_t GetProtocolId() const { return mProtocolId; }
+
+private:
+    constexpr uint32_t ToUint32() const { return (static_cast<uint32_t>(mVendorId) << 16) | mProtocolId; }
+
+    chip::VendorId mVendorId;
+    uint16_t mProtocolId;
 };
 
+// Common Protocols
+//
+// NOTE: Do not attempt to allocate these values yourself.
+#define CHIP_STANDARD_PROTOCOL(name, id)                                                                                           \
```

Comment:
nit: I wonder if there would be a way to skip the macro. I was going to propose to name things like `chip::Protocol::SecureChannel` however I see that SecureChannel/Echo/etc names are already used.
Maybe since these are constexpr we could use the `k*` naming convention:

```cpp
static constexpr Protocols::Id kSecureChannel(VendorId::Common, SecureChannel);
static constexpr Protocols::Id kEcho(VendorId::Common, Echo);
//...
```

Reply:
Right, the macro is just there because it's more readable, arguably, than the many namespace blocks.

And yes, `chip::Protocols::SecureChannel` is already a namespace.

We could absolutely do `chip::Protocols::kSecureChannel` or `chip::Protocols::SecureChannel_Id` or `chip::ProtocolIds::SecureChannel` or something like that.  I was pretty torn on the naming, and don't have particularly strong views about it in the end.  Something inside `chip::Protocols` is probably easier to use in code that is already in that namespace or pulls it in with `using` than `chip::ProtocolIds` is....

Reply:
Oh, another option is to move the id definitions into the relevant namespaces or whatnot in the per-protocol header files.  But then you'd need to include the SecureChannel constants header to use the SecureChannel id, which may not be desirable.  And it would be harder to see the full set of protocol ids and their values all at once, which is not great either.